protocols/perf/src/client/handler.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{
+    collections::VecDeque,
+    task::{Context, Poll},
+};
+
+use futures::{
+    stream::{BoxStream, SelectAll},
+    StreamExt,
+};
+use libp2p_core::upgrade::{DeniedUpgrade, ReadyUpgrade};
+use libp2p_swarm::{
+    handler::{
+        ConnectionEvent, DialUpgradeError, FullyNegotiatedInbound, FullyNegotiatedOutbound,
+        ListenUpgradeError,
+    },
+    ConnectionHandler, ConnectionHandlerEvent, KeepAlive, StreamProtocol, SubstreamProtocol,
+};
+use void::Void;
+
+use super::{RunError, RunId};
+use crate::{RunParams, RunUpdate};
+
+#[derive(Debug)]
+pub struct Command {
+    pub(crate) id: RunId,
+    pub(crate) params: RunParams,
+}
+
+#[derive(Debug)]
+pub struct Event {
+    pub(crate) id: RunId,
+    pub(crate) result: Result<RunUpdate, RunError>,
+}
+
+pub struct Handler {
+    /// Queue of events to return when polled.
+    queued_events: VecDeque<
+        ConnectionHandlerEvent<
+            <Self as ConnectionHandler>::OutboundProtocol,
+            <Self as ConnectionHandler>::OutboundOpenInfo,
+            <Self as ConnectionHandler>::ToBehaviour,
+            <Self as ConnectionHandler>::Error,
+        >,
+    >,
+
+    requested_streams: VecDeque<Command>,
+
+    outbound: SelectAll<BoxStream<'static, (RunId, Result<crate::RunUpdate, std::io::Error>)>>,
```

Comment:
Perhaps use `futures_bounded` here now?

Reply:
That one does all the waker stuff correctly for you. `SelectAll` returns `None` when it is empty so technically, you'd have to register a waker.

Reply:
Today `futures-bounded` only offers abstractions for `Future`s and not `Stream`s. Please correct me if I am wrong.

Happy to add such abstraction. Though I don't want to block this pull request on it.

Reply:
https://github.com/libp2p/rust-libp2p/pull/4616 :)

Reply:
> Today `futures-bounded` only offers abstractions for `Future`s and not `Stream`s. Please correct me if I am wrong.

You are not completely wrong but also not completely right :)

See https://github.com/libp2p/rust-libp2p/pull/4616.

Reply:
:D